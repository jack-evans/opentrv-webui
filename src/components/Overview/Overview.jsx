import React, { Component } from 'react'
import OverviewHeader from '../OverviewHeader/OverviewHeader'
import { Button, Loading, Search } from 'carbon-components-react'
import DeviceTile from '../DeviceTile/DeviceTile'
import axios from 'axios'

class Overview extends Component {
  constructor (props) {
    super(props)
    this.state = {
      devices: [],
      visibleDevices: [],
      isLoading: true
    }
    this.search = ''
    this.timer = {}

    this.discoverDevicesButtonOnClick = this.discoverDevicesButtonOnClickHandler.bind(this)
    this.searchOnChange = this.searchOnChangeHandler.bind(this)
    this.searchDevices = this.filterDevices.bind(this)
  }

  componentDidMount () {
    const tick = () => {
      return Overview.retrieveDevices()
        .then(devices => {
          this.setState({
            devices: devices,
            visibleDevices: devices.filter(this.searchDevices),
            isLoading: false
          })

          this.timer = setTimeout(tick, 1000)
        })
    }
    tick()
  }

  componentWillUnmount () {
    clearTimeout(this.timer)
  }

  searchOnChangeHandler (event) {
    this.search = event.target.value
    this.setState({visibleDevices: this.state.devices.filter(this.searchDevices)})
  }

  filterDevices (device) {
    const searchTerm = this.search.toLowerCase()

    return (
      (device.doc.name.toLowerCase().indexOf(searchTerm) !== -1) ||
      (device.doc.currentTemperature.toString().indexOf(searchTerm) !== -1)
    )
  }

  async discoverDevicesButtonOnClickHandler () {
    let retrievedDevices = await Overview.discoverDevices('yes')
    this.setState({
      devices: retrievedDevices,
      visibleDevices: retrievedDevices
    })
  }

  static retrieveDevices () {
    return Overview.discoverDevices('no')
  }

  static discoverDevices (userFlag) {
    let options = {
      url: `/api/v1/devices?user=${userFlag}`,
      method: 'GET',
      json: true
    }

    return axios(options)
      .then(response => response.data)
  }

  deleteDevice (id) {
    return axios({
      url: `/api/v1/devices/${id}`,
      method: 'DELETE',
      json: 'true'
    }).then(() => {
      return Overview.discoverDevices('no')
    }).then(retrievedDevices => {
      this.setState({
        devices: retrievedDevices,
        visibleDevices: retrievedDevices
      })
    })
  }

  /**
   * render method
   *
   * renders the overview component of the UI experience
   * @returns {HTML} - Overview component
   */
  render () {
    let overviewContent
    let overviewContentClass = 'Overview__content'

    if (this.state.isLoading) {
      overviewContentClass += ' Overview__content-loading'
      overviewContent = (
        <div>
          <div style={{'display': 'flex', 'justifyContent': 'center'}}>
            <Loading small withOverlay={false} />
          </div>
          <div>
            Loading. Please Wait...
          </div>
        </div>
      )
    } else if (this.state.devices.length < 1) {
      overviewContentClass += ' Overview__content-empty'
      overviewContent = (
        <div>
          <div className='Overview__content-empty-message'>
            <h3>You currently do not have any devices registered</h3>
          </div>
          <div>
            <Button
              className='Overview__discover_button'
              icon='add--outline'
              iconDescription='Find devices that are on your network'
              onClick={this.discoverDevicesButtonOnClick}
            >
              Discover my devices
            </Button>
          </div>
        </div>
      )
    } else {
      // This is where the tiles for each device are build up
      overviewContent = this.state.visibleDevices.map(device => {
        return (
          <DeviceTile
            key={device.doc.id}
            id={device.doc.id}
            name={device.doc.name}
            temp={device.doc.currentTemperature}
            active={device.doc.active}
            handleDelete={() => { this.deleteDevice(device.doc.id) }}
          />
        )
      })
    }

    return (
      <div className='Overview'>
        <OverviewHeader />
        <div className='Overview__search-bar-container'>
          <div>
            <Search
              className='Overview__search-bar'
              small
              id='device-search'
              labelText='Search'
              placeHolderText='Search Devices'
              onChange={this.searchOnChange}
            />
          </div>
        </div>
        <div className={overviewContentClass}>
          {overviewContent}
        </div>
      </div>
    )
  }
}

export default Overview
