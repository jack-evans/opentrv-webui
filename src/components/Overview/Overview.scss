
.Overview {

  .Overview__search-bar-container {
    border-bottom: 2px solid $ui-04;
    display: flex;
    height: 100px;
    justify-content: center;
    padding: 30px 0 30px 0;

    .Overview__search-bar {
      width: 500px;
    }
  }

  .Overview__content {
    display: flex;
    flex-wrap: wrap;
    padding-bottom: 20px;
  }

  .Overview__content-loading {
    align-content: center;
    justify-content: center;
    padding-top: 10px;
    text-align: center;
  }

  .Overview__content-empty {
    align-content: center;
    flex: 1 1 auto;
    flex-flow: column;
    height: calc(100vh - 260px);
    justify-content: center;

    .Overview__content-empty-message {
      padding-bottom: 10px;
    }
  }
}
