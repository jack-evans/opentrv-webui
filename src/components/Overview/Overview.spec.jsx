import React from 'react'
import Overview from './Overview'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

Enzyme.configure({ adapter: new Adapter() })

describe('Overview.jsx', () => {
  it('renders without crashing', () => {
    expect(shallow(<Overview />)).toMatchSnapshot()
  })

  describe('when the tabSelected state is 0', () => {
    it('renders the DeviceOverview', () => {
      const wrapper = shallow(<Overview />)
      wrapper.setState({tabSelected: 0})
      expect(wrapper.find('DeviceOverview')).toHaveLength(1)
    })
  })

  describe('when the tabSelected state is 1', () => {
    it('renders the PolicyOverview', () => {
      const wrapper = shallow(<Overview />)
      wrapper.setState({tabSelected: 1})
      expect(wrapper.find('PolicyOverview')).toHaveLength(1)
    })
  })

  describe('when the tabSelected state is another number', () => {
    it('renders null', () => {
      const wrapper = shallow(<Overview />)
      wrapper.setState({tabSelected: 1000})
      expect(wrapper.find('PolicyOverview')).toHaveLength(0)
    })
  })
})
