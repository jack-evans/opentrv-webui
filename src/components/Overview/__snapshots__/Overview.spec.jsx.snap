// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview.jsx renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Overview />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <OverviewHeader />,
        <div
          className="Overview__search-bar-container"
        >
          <div>
            <Search
              className="Overview__search-bar"
              id="device-search"
              labelText="Search"
              onChange={[Function]}
              placeHolderText="Search Devices"
              small={true}
              type="text"
            />
          </div>
        </div>,
        <div
          className="Overview__content"
        >
          <div>
            <DeviceTile
              active={true}
              name="Device 4"
              temp={25.2}
            />
          </div>
        </div>,
      ],
      "className": "Overview",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            <Search
              className="Overview__search-bar"
              id="device-search"
              labelText="Search"
              onChange={[Function]}
              placeHolderText="Search Devices"
              small={true}
              type="text"
            />
          </div>,
          "className": "Overview__search-bar-container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Search
              className="Overview__search-bar"
              id="device-search"
              labelText="Search"
              onChange={[Function]}
              placeHolderText="Search Devices"
              small={true}
              type="text"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": "Overview__search-bar",
              "id": "device-search",
              "labelText": "Search",
              "onChange": [Function],
              "placeHolderText": "Search Devices",
              "small": true,
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            <DeviceTile
              active={true}
              name="Device 4"
              temp={25.2}
            />
          </div>,
          "className": "Overview__content",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <DeviceTile
              active={true}
              name="Device 4"
              temp={25.2}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "active": true,
              "name": "Device 4",
              "temp": 25.2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <OverviewHeader />,
          <div
            className="Overview__search-bar-container"
          >
            <div>
              <Search
                className="Overview__search-bar"
                id="device-search"
                labelText="Search"
                onChange={[Function]}
                placeHolderText="Search Devices"
                small={true}
                type="text"
              />
            </div>
          </div>,
          <div
            className="Overview__content"
          >
            <div>
              <DeviceTile
                active={true}
                name="Device 4"
                temp={25.2}
              />
            </div>
          </div>,
        ],
        "className": "Overview",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              <Search
                className="Overview__search-bar"
                id="device-search"
                labelText="Search"
                onChange={[Function]}
                placeHolderText="Search Devices"
                small={true}
                type="text"
              />
            </div>,
            "className": "Overview__search-bar-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Search
                className="Overview__search-bar"
                id="device-search"
                labelText="Search"
                onChange={[Function]}
                placeHolderText="Search Devices"
                small={true}
                type="text"
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": "Overview__search-bar",
                "id": "device-search",
                "labelText": "Search",
                "onChange": [Function],
                "placeHolderText": "Search Devices",
                "small": true,
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              <DeviceTile
                active={true}
                name="Device 4"
                temp={25.2}
              />
            </div>,
            "className": "Overview__content",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <DeviceTile
                active={true}
                name="Device 4"
                temp={25.2}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "active": true,
                "name": "Device 4",
                "temp": 25.2,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
